### PR TITLE
feat(react-instantsearch): add `Carousel` layout component

### PR DESCRIPTION
**Pull request by shaejaz**
This adds the the `Carousel` layout component and enables the `layoutComponent` prop in Recommend widgets. React getting started example has been update to use `Carousel` as well.

image

https://algolia.atlassian.net/browse/FX-2819

### CODE REVIEW

File: packages/react-instantsearch/src/templates/index.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './Carousel';
```
Comment (Haroenv) [inline packages/react-instantsearch/src/templates/index.ts:1]: it's worth thinking about, but in vanilla it's called /templates because that exists as a concept already, in react the components end in Component, so they should be in a components folder I think (or widgets is probably fine too, Highlight isn't a widget per se either and is exported from widgets)

File: examples/react/getting-started/src/App.css [hunk 2]
```diff
@@ -61,7 +61,7 @@ em {
   flex: 3;
 }
 
-.searchbox {
+#searchbox {
```
Comment (sarahdayan) [inline examples/react/getting-started/src/App.css:64]: Why this change? I don't see any element with `id="searchbox"` in the example.

File: packages/react-instantsearch/src/templates/Carousel.tsx
Comment (sarahdayan) [inline packages/react-instantsearch/src/templates/Carousel.tsx:1]: I'd rename this folder `components`. React InstantSearch doesn't have the concept of "templates", and the component props are all called `somethingComponent`.

File: packages/react-instantsearch/src/widgets/__tests__/FrequentlyBoughtTogether.test.tsx [hunk 3]
```diff
@@ -81,4 +82,149 @@ describe('FrequentlyBoughtTogether', () => {
...
+    });
+    const { container } = render(
+      <InstantSearchTestWrapper searchClient={client}>
+        <FrequentlyBoughtTogether
+          objectIDs={['1']}
+          itemComponent={({ item }) => <p>{item.objectID}</p>}
+          layoutComponent={(props) => (
+            <Carousel
+              {...props}
+              previousIconComponent={() => <p>Previous</p>}
```
Comment (sarahdayan) [inline packages/react-instantsearch/src/widgets/__tests__/FrequentlyBoughtTogether.test.tsx:151]: Can you also test passing custom `classNames` and `translations` since they're available APIs?

Same for the other widgets.

File: packages/instantsearch-ui-components/src/types/Recommend.ts
Comment (sarahdayan) [inline packages/instantsearch-ui-components/src/types/Recommend.ts:1]: Correct change IMO, but we're still passing `translations` to the layout in each UI component.

**Review by sarahdayan — CHANGES_REQUESTED** [review]
Almost there, a couple changes and we should be good.

I also noticed that the box shadow is being cut off in demo because of the overflow ruleset. I fixed it in the JS example, feel free to compare markup and styles to replicate it.

Capture d’écran 2024-08-09 à 15 21 25

File: packages/react-instantsearch/src/templates/index.ts (hunk 1 above)
Comment (sarahdayan) [reply packages/react-instantsearch/src/templates/index.ts:1]: This or `layout`. It doesn't change anything for users who will import from `react-instantsearch`, but internally it makes more sense.

File: examples/react/getting-started/src/App.css [hunk 4]
```diff
@@ -113,3 +107,50 @@ em {
   height: 100%;
   justify-content: space-between;
 }
+
+.ais-TrendingItems-item h2,
+.ais-RelatedProducts-item h2 {
+  display: -webkit-box;
```
Comment (Haroenv) [inline examples/react/getting-started/src/App.css:113]: where did this style come from? seems copy-pasted from somewhere with autoprefixer maybe?

File: packages/instantsearch-ui-components/src/components/TrendingItems.tsx [hunk 5]
```diff
@@ -90,7 +90,6 @@ export function createTrendingItemsComponent({
 
         <Layout
           classNames={cssClasses}
-          translations={translations}
```
Comment (Haroenv) [inline packages/instantsearch-ui-components/src/components/TrendingItems.tsx:93]: does this mean the "previous" and "next" labels aren't customisable, or is that done another way now?

**Review by Haroenv — COMMENTED** [review]
🔥

File: examples/react/getting-started/src/App.css (hunk 4 above)
Comment (sarahdayan) [reply examples/react/getting-started/src/App.css:113]: Same styles as the ones I used for InstantSearch.js. They're the recommended ones for line clipping.

File: packages/react-instantsearch/src/widgets/__tests__/__utils__/cx.ts
Comment (shaejaz) [inline packages/react-instantsearch/src/widgets/__tests__/__utils__/cx.ts:1]: Not sure if this should be here, it's only being used by the updated tests in the react-is package atm.

Comment (Haroenv) [reply packages/react-instantsearch/src/widgets/__tests__/__utils__/cx.ts:1]: I think the correct way is to `import { cx } from 'instantsearch-ui-components';`

File: packages/instantsearch-ui-components/src/components/TrendingItems.tsx (hunk 5 above)
Comment (sarahdayan) [reply packages/instantsearch-ui-components/src/components/TrendingItems.tsx:93]: The translations we passed were the ones of the widget, and it didn't make sense to do so. Layouts have their own translations which they receive as a prop.

**Review by Haroenv — APPROVED** [review]

**Review by dhayab — APPROVED** [review]
👏

**Review by sarahdayan — APPROVED** [review]

**Review by Haroenv — APPROVED** [review]
